doc/developer/design/20210406_mapfilterproject.md

```diff
@@ -0,0 +1,155 @@
+
+The `MirRelationExpr::Map` operator contains a sequence of `MirScalarExpr` expressions.
+Each expression is a function that is applied to the existing columns, and which produces a new column.
+The sequence of expressions indicates a sequence of new columns to add to each row input to the `Map`.
+
+The `MirRelationExpr::Filter` operator contains a sequence of `MirScalarExpr` expressions.
+Each expression is a function that is applied to the existing columns, and which produces either `Datum::True` or not.
+Rows that produce `Datum::True` are retained in the output, and rows that do not are discarded.
+
+The `MirRelationExpr::Project` operator contains a sequence of integer column references.
+Each column reference indicates that the value found in that column should be copied to the output in place of the reference.
+The projection allows us to remove various columns from the input, and re-order and duplicate existing columns.
+
+These three operations can by applied row-by-row, as their actions have no relation to other rows in the system.
+Arbitrary sequences of these operations can always be reduced to one `Map`, then one `Filter`, then one `Project`.
+
+The proposal is to assemble and communicate this information (caveat: some modifications will be made) to sources:
+```rust
+pub struct MapFilterProject {
+    /// A sequence of expressions that should be appended to the row.
+    pub expressions: Vec<MirScalarExpr>,
+    /// Expressions that must evaluate to `Datum::True` for the output
+    /// row to be produced.
+    pub predicates: Vec<MirScalarExpr>,
+    /// A sequence of column identifiers whose data form the output row.
+    pub projection: Vec<usize>,
+}
+```
+The three members correspond exactly to one `Map`, then one `Filter`, then one `Project`.
+
+On instantiation each source would be supplied with a `&mut MapFilterProject` argument.
+Depending on the capabilities of the source, it could take any or all of the operations, and leave behind any work undone.
+A simple default would be to leave the argument untouched, communicating that it applied none of the operations.
+A thorough source could take the entire instance and leave behind an identity operator, communicating that there is no remaining work to be done.
+
+### Manipulation
+
+The `MapFilterProject` structure has several methods that are meant to facilitate its manipulation.
+For example, the `demand(&self)` method indicates which input columns will be examined by the instance.
+Additionally, the `permute(&mut self, ..)` method re-writes all column references as if the input had been permuted.
+These two together allow sources to determine which of their announced columns need to be constructed at all, and to rewrite a `MapFilterProject` instance to reference only those columns that were constructed.
+
+As an example, the `csv.rs` source provides the appearance of multiple columns of text.
```

Comment:
So I want to make sure I understand the example correctly.

If a csv source has 7 columns but we only use the last three (so the mfp just projects out everything but columns 4, 5, and 6), we could look at the demand information and only produce those columns (now at positions 0, 1, 2) and `permute` would do the work to make the MFP handle the new column mappings and make it become an identity?

Reply:
I think I would benefit from a slightly larger example here, like if a simple mfp said "filter out everything where column 2 < 0" and my source knew how to do that before decoding the input into rows ... how would I express to the mfp that I already did the filtering? or would i not do that in that case?